### PR TITLE
BUG: handle deprecation warning from numpy 1.25 (np.product is deprecated in favour of np.prod)

### PR DESCRIPTION
**Pull request by neutrinoceros**
In repsonse to https://github.com/numpy/numpy/pull/23314

### CODE REVIEW

**Comment by takluyver** [comment]
Thanks! While we're looking at this, though, I think all or most of these occurrences can be replaced by our internal `product` function. We found at some point that for a handful of integers, as in a shape tuple, it's quicker to multiply them in Python than to get Numpy involved.

https://github.com/h5py/h5py/blob/7705f672cc6e091b069900ea260c3a27f329445a/h5py/_hl/base.py#L504-L513

**Comment by neutrinoceros** [comment]
@takluyver cool ! I've replaced every occurrences where `np.product` was used on an array shape.

File: h5py/_hl/filters.py
Comment (tacaswell) [inline h5py/_hl/filters.py:361]: I would move these over too?  I think chunks are still shape-like.

Not going to block merging over.

**Review by tacaswell — APPROVED** [review]

File: h5py/_hl/filters.py
Comment (neutrinoceros) [reply h5py/_hl/filters.py:361]: Your call, I just didn't want to optimize code that I didn't immediately understand :)

Comment (neutrinoceros) [reply h5py/_hl/filters.py]: done ! FWIW, I measured that `h5py._hl.base.product` was indeed faster than `np.prod` on small arrays such as `chunks`

**Comment by takluyver** [comment]
Thanks @neutrinoceros !

**Comment by pllim** [comment]
Hello. When will this patch be in a release? I see this warning now with numpy 1.25rc1. Thank you!

**Pull request merged**
